### DISCUSSION TITLE
Use Remote-User header also for Subonic API requests

### DISCUSSION THREAD

**Discussion by crazygolem** [navidrome/navidrome#2487] [Ideas]
I'm evaluating options to address the security issues with the subsonic API's authentication scheme and what it implies for me as the service provider; namely I don't like that user-defined passwords get stored encrypted instead of hashed in my database.

Currently, I am using the reverse proxy authentication mechanism to delegate it to another component (so I can have other services using the same credentials behind the same proxy), but this explicitly doesn't work with navidrome for the subsonic API, so I just chose to block the subsonic endpoint entirely.

Because of this, I entirely disabled the user self-management in navidrome, which also has the benefit of not confusing my not-so-tech-savvy users if they went to change their password in the wrong place and nothing happened.

I don't see why the subsonic authentication couldn't also be delegated to a dedicated authentication service (which could still perform subsonic's standard authentication, so would be entirely transparent to clients).

For me the advantage would be:
- I can implement a credential management scheme I'm more confortable with, namely I'm thinking of not letting users provide their passwords and instead generating them (essentially using API keys).
- To manage all authentication and credentials in the same place, with the possibility of providing a streamlined and consistent user experience for credential management: it's easier to explain to my family members they have to go to the same place for any kind of credentials, instead of needing to explain which credential has to be changed where for what purpose.

I think this could also realistically be a first step to get subsonic clients to experiment with better authentication schemes, if they know that there is at least some servers where it could be integrated without having to update the music server itself.

@deluan would you be against the possibility of using the reverse proxy authentication's `Remote-User` header also for the subsonic requests?

Please note that I'm not proposing to address the encrypted-not-hashed-password issue at this point, just to let navidrome delegate this mess to another specialized component.

**Comment by crazygolem** [comment]
To get the ball rolling, I implemented a [Traefik plugin](https://github.com/crazygolem/traefik-subsonic-basicauth) to create a BasicAuth header based on the SubsonicAuth parameters.
This will allow to transparently integrate subsonic clients that don't already support BasicAuth, and let a third-party service perform the authentication.

I also have a branch of Navidrome ready to accept the reverse proxy's user header for the Subsonic endpoint, and I plan to try it out during the week if I have some time.

**Comment by slashformotion** [comment]
If  I understand navidrome code correctly, when a user is created and there is an encryption key setup by the admin of the navidrome instance, all passwords encrypted symmetrically using AES. Why not use the "usual" SALT+HASH method ? 

I am somewhat confused

**Comment by deluan** [comment]
Navidrome needs to be able to decrypt to clear text the original password. It is a requirement for Subsonic API authentication. See discussions in #202 and https://github.com/airsonic/airsonic/issues/69

**Reply by slashformotion** [reply]
Ok, I got it. I looked at the docs from subsonic 

(for futur references: http://www.subsonic.org/pages/api.jsp#versions)
<img width="965" alt="image" src="https://github.com/navidrome/navidrome/assets/45801817/398d4aa2-eb1a-4e4b-8084-f3fff07f55ed">

I mean, it's a strange choice

**Reply by slashformotion** [reply]
> "I can implement a credential management scheme I'm more confortable with, namely I'm thinking of not letting users provide their passwords and instead generating them (essentially using API keys)."

This should do the trick
